UnicornRender/video/source/Camera.cpp

```diff
@@ -12,120 +12,290 @@ namespace unicorn
 {
 namespace video
 {
-Camera::Camera(const glm::vec3& position, const glm::vec3& direction) : 
-    m_aspect(0)
-    , m_position(position)
-    , m_upVector(0.0f, -1.0f, 0.0f)
-    , m_direction(direction)
-    , m_fov(45.0f)
-    , m_znear(0.1f)
-    , m_zfar(1000.0f)
-    , m_dirtyView(false)
-    , m_fovLowerBound(44.0f)
-    , m_fovUpperBound(45.0f)
-    , m_dirtyProjection(false)
-{
-    UpdateViewMatrix();
-    UpdateProjectionMatrix();
-}
-
-void Camera::SetPerspective(float fov, float aspect, float znear, float zfar)
-{
-    m_fov = fov;
-    m_znear = znear;
-    m_zfar = zfar;
-    m_aspect = aspect;
-    m_dirtyProjection = true;
+
+void CameraController::SetDirection(const glm::vec3& direction)
+{
+    m_direction = direction;
+    m_rightVector = glm::normalize(glm::cross(m_direction, m_upVector));
+    m_isDirty = true;
 }
 
-void Camera::SetAspectRatio(float aspect)
+void CameraController::SetUpVector(const glm::vec3& upVector)
 {
-    m_aspect = aspect;
-    m_dirtyProjection = true;
+    m_upVector = upVector; 
+    m_rightVector = glm::normalize(glm::cross(m_direction, m_upVector));
+    m_isDirty = true;
 }
 
-void Camera::Translate(const glm::vec3& delta)
+void CameraController::SetPosition(const glm::vec3& position)
 {
-    m_position += delta;
-    m_dirtyView = true;
+    m_position = position;
+    m_isDirty = true;
 }
 
-void Camera::SetDirection(const glm::vec3& direction)
+glm::vec3 const& CameraController::GetDirection() const
 {
-    m_direction = direction;
-    m_dirtyView = true;
+    return m_direction;
 }
 
-void Camera::SetUpVector(const glm::vec3& upVector)
+glm::vec3 const& CameraController::GetUpVector() const
 {
-    m_upVector = upVector;
-    m_dirtyView = true;
+    return m_upVector;
 }
 
-void Camera::SetPosition(const glm::vec3& position)
+glm::vec3 const& CameraController::GetPosition() const
 {
-    m_position = position;
-    m_dirtyView = true;
+    return m_position;
 }
 
-const glm::vec3& Camera::GetDirection() const
+void CameraController::Recalculate()
 {
-    return m_direction;
+    if (m_isDirty)
+    {
+        UpdateViewMatrix();
+        m_isDirty = false;
+    }
 }
 
-const glm::vec3& Camera::GetUpVector() const
+CameraController::CameraController(glm::mat4& cameraView) :
+    m_cameraView(cameraView),
+    m_position(0.0),
+    m_upVector(0.0f, -1.0f, 0.0f),
+    m_direction(0.0),
+    m_rightVector(glm::cross(m_direction, m_upVector)),
+    m_speed(100.f),
```

Comment:
what is the point of `m_speed` if it's not configurable? I believe that interface for `MoveX` will be cleaner if you will be operating just in `float`s of distance instead